src/functions/defaultsPure.js

```diff
@@ -2,14 +2,32 @@
 
 // NOTE: this behaves like lodash/defaults, but doesn't mutate the target
 // it also preserve keys order
+var PLACEHOLDER = '__PLACEHOLDER__';
+
 module.exports = function defaultsPure() {
   var sources = Array.prototype.slice.call(arguments);
+
+  var emptyObjectWithKeysInOrder = sources.reduceRight(function(acc, source) {
+    Object.keys(Object(source)).forEach(function(key) {
+      if (source[key] === undefined) {
+        return;
+      }
+      if (acc[key] === PLACEHOLDER) {
+        // remove if already added, so that we can add it in correct order
+        delete acc[key];
+      }
+      acc[key] = PLACEHOLDER;
+    });
+    return acc;
+  }, {});
+
   return sources.reduce(function(acc, source) {
     Object.keys(Object(source)).forEach(function(key) {
-      if (source[key] !== undefined && !Object.hasOwnProperty.call(acc, key)) {
+      if (source[key] !== undefined && acc[key] === PLACEHOLDER) {
         acc[key] = source[key];
       }
     });
     return acc;
-  }, {});
+  }, emptyObjectWithKeysInOrder);
```

Comment:
Maybe you can skip the `emptyObjectWithKeysInOrder` step by just checking if the key already exists:

```js
module.exports = function defaultsPure() {
  var sources = Array.prototype.slice.call(arguments);

  return sources.reduceRight(function(acc, source) {
    Object.keys(Object(source)).forEach(function(key) {
      if (source[key] === undefined) {
        return;
      }
      if (acc[key] !== undefined) {
        // remove if already added, so that we can add it in correct order
        delete acc[key];
      }
      acc[key] = source[key];
    });
    return acc;
  }, {});
};
```

WDYT ?

Reply:
To merge the values, we go from left to right. But to preserve the order of keys, we should go from right to left. That's why I had to have the separate step. And without it, I couldn't think of a way to preserve the order just in one reducer by checking existence of keys. Any idea?

Reply:
Mmm, the test suite is passing with this code. If this is incorrect maybe we are missing a test case here ?

Reply:
Oh you're right. I misunderstood your code. It works fine. Great job!!! https://github.com/algolia/algoliasearch-helper-js/pull/786/commits/7b88a74908a35c2a99531c3a1b9696c3dce254ee